### PR TITLE
auto detect conflicting tags

### LINKED ISSUE #111: Find conflicting fields programatically
We embed a lot of structs (huuuge structs), and it's very likely that there are some conflicting fields which we cannot catch just by looking.

We need to way to find conflicting fields in our resulting spec, and then handle such conflicts gracefully and very reliably.

One example is while implementing jobs (#52), I figured that the `activeDeadlineSeconds` field would not populate in the JobSpec, and @surajssd and I figured out that this field exists in both `api.PodSpec` and `batch.JobSpec` and the way we are building jobs is -

```golang
type JobSpecMod struct {
	Name          string `json:"name,omitempty"`
	api.PodSpec   `json:",inline"`
	batch.JobSpec `json:",inline"`
}
```

Hence, this field does not get populated in either of the structs due to the conflict.

In this case, the conflicting field `activeDeadlineSeconds` specifies the number of seconds a pod is allowed to run on a node, so it makes sense to substitute the same value in both `JobSpec` and `PodSpec`.

Similarly, we need to take decisions for all the conflicting fields.

So -
- [ ] Find conflicting fields
- [ ] Handle conflicting fields such that the unmarshalled data for these is not lost and is stored somewhere
- [ ] Make intelligent decision for each conflicting field based on the input data

### PR DESCRIPTION
**Pull request by containscafeine**
- [x] pass all structs besides spec.App
- [x] move everything to a test
- [x] add comments
- [x] add tests for the behavior
- [x] take care of pointer mechanics all over

EDIT:

Before this commit, we had no way to know if the embedded structs
had any conflicting fields between each other or with the struct
embedding the structs. This led to inconsistency since in case
of conflicting fields, the data would not be unmarshalled and due
to presence of a lot of fields, this would go unnoticed.

This commit handles this in a 4 step process.

Step 1: Get all the target structs.
Look at tags in the fields of the input struct as well as the embedded
structs, so both of them become our target structs.

Step 2: Get blacklisted tags.
Get the fields which have already been handled and were previously
conflicting. This needs to be done only for the structs in spec.go,
and with fields marked as "conflicting" using a JSON tag

Step 3: Get JSON tags from target structs.
Get the JSON tags from all of the target structs which we got in Step 1
except for the blacklisted tags, along with the names of the structs
which hold the field with that tag.

Step 4: Check if more than one struct holds a tag.
Delete all the tags in Step 3 which are only help by one struct, because
it means that the JSON tag is not conficting, and finally return the
remaining tags which are held by more than one struct, and are not
handled, hence conflicting tags

The code has been heavily documented, and tests have been added for this
behavior.

The entire code is not a part of the main program and is not run every
time a user executes any kedge command, instead it has been put as a
test, since it's more of a developer facing problem than a user facing
one.

Fixes #111

### CODE REVIEW

**Comment by kadel** [comment]
is this related to https://github.com/kedgeproject/kedge/issues/111 ?

**Comment by containscafeine** [comment]
@kadel yep, #135 fixes #111

**Comment by containscafeine** [comment]
PTAL @surajssd @cdrage @kadel

File: pkg/spec/spec.go [hunk 1]
```diff
@@ -37,7 +37,7 @@ type ServicePortMod struct {
 type ServiceSpecMod struct {
 	api_v1.ServiceSpec `json:",inline"`
 	Name               string           `json:"name,omitempty"`
-	Ports              []ServicePortMod `json:"ports"`
+	Ports              []ServicePortMod `json:"ports,conflicting"`
```
Comment (surajssd) [inline pkg/spec/spec.go]: (not read code yet) what does this json tag mean? this should be automatically detected right?

Comment (containscafeine) [inline pkg/spec/spec.go]: No, we need to add this JSON tag to every conflicting tag that we have handled.
e.g. Ports are conflicting in the sense that they exist in SeviceSpec as well as in our spec in ServiceSpecMod, so we need to declare it as conflicting meaning that this is a known conflicting tag and it has been handled.

File: pkg/spec/util.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+			}
+		}
+	}
+	return blacklistedTags, nil
+}
+
+// This function checks if the input parameter has the pointer to a struct as
+// its underlying type.
+// Returns "nil" if true, or an error if false
+func isPointerToStruct(input interface{}) error {
```
Comment (surajssd) [inline pkg/spec/util.go]: nit: ideally functions with `isFooBar` types should return `boolean`

because the error we are getting does not matter to us in a way we just wanna check if it is struct or not.

But also makes sense to return error because it helps us understand what are we passing!

File: pkg/spec/spec.go (hunk 1 above)
Comment (surajssd) [inline pkg/spec/spec.go]: Maybe add a comment above that field saying such!

File: pkg/spec/util.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// This function takes an map as an input which stores the JSON tag as key and
+// the names of the structs holding that tag as the value. The passed in
+// blaclisted tags are ignored and are not populated at all.
```
Comment (surajssd) [inline pkg/spec/util.go]: `blacklisted*`

File: pkg/spec/util.go [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+		field := val.Type().Field(i)
+
+		// We do not need to store JSON tags like "inline" or "omitempty",
+		// instead we only need JSON tag which is used for unmarshalling the
+		// YAML. It's safe to assume that, that JSON tag is the one that
+		// is specified first.
+		// Let's split the JSON tag on ",", and store the first element of the
+		// resulting array.
+
+		unmarshalJsonTag := strings.SplitN(field.Tag.Get("json"), ",", 2)[0]
```
Comment (surajssd) [inline pkg/spec/util.go]: you can use the library https://github.com/fatih/structtag this helps you parse json tags

I have used it when working on `spec.go` parsing here as:

[`structtag.Parse(j)`](https://github.com/surajssd/kedgeSchema/blob/17adbe3d19a6a995cb957b11bd7aa159a558dd6b/parsego.go#L338)

It basically parses string and gives you list of all all tags, giving an index like this might cause panics later when we encounter the use cases which we are not thinking now!

File: pkg/spec/util.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+		skipTag := false
+		// Populating only the non blacklisted tags
+		for _, blacklistedTag := range blacklistedTags {
+			if unmarshalJsonTag == blacklistedTag {
+				skipTag = true
```
Comment (surajssd) [inline pkg/spec/util.go]: once this is set to true you can safely break out of for loop

File: pkg/spec/util.go [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+			}
+		}
+		if unmarshalJsonTag == "" {
+			// It does not make any sense to store anything for JSON tags with no
+			// first element, e.g. `json:",inline"`
+			skipTag = true
```
Comment (surajssd) [inline pkg/spec/util.go]: if `skipTag` decides whether to skip the loop, then we can just put `continue` statement here.

File: pkg/spec/util.go [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+		unmarshalJsonTag := strings.SplitN(field.Tag.Get("json"), ",", 2)[0]
+
+		skipTag := false
+		// Populating only the non blacklisted tags
+		for _, blacklistedTag := range blacklistedTags {
+			if unmarshalJsonTag == blacklistedTag {
+				skipTag = true
+			}
+		}
+		if unmarshalJsonTag == "" {
```
Comment (surajssd) [inline pkg/spec/util.go]: if this is checked before the `blacklist` for loop we can reduce some looping

File: pkg/spec/spec.go (hunk 1 above)
Comment (containscafeine) [inline pkg/spec/spec.go]: Above every conflicting field? I'd rather want this to be a part of the developer docs, and the tests are going to fail anyway if a field is not handled and is not marked as "conflicting".

File: pkg/spec/util.go [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// This function takes an map as an input which stores the JSON tag as key and
```
Comment (containscafeine) [inline pkg/spec/util.go]: fixed

File: pkg/spec/util.go (hunk 3 above)
Comment (containscafeine) [inline pkg/spec/util.go]: fixed

File: pkg/spec/util.go (hunk 4 above)
Comment (containscafeine) [inline pkg/spec/util.go]: Umm, not sure if we need to use an external library for our use case.
And the panics should be handled by our tests.

File: pkg/spec/util.go (hunk 5 above)
Comment (containscafeine) [inline pkg/spec/util.go]: makes sense, refactored

File: pkg/spec/util.go (hunk 7 above)
Comment (containscafeine) [inline pkg/spec/util.go]: makes sense, refactored

File: pkg/spec/util.go (hunk 6 above)
Comment (containscafeine) [inline pkg/spec/util.go]: makes sense, refactored

File: pkg/spec/util.go (hunk 7 above)
Comment (containscafeine) [inline pkg/spec/util.go]: makes sense, refactored

File: pkg/spec/util.go [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+	}
+	return embeddedStructs, nil
+}
+
+// This function returns an array of all the JSON tags used for unmarshalling
+// (which are the first ones specified) which have a JSON tag "conflicting", in
+// all of the input structs.
+// It's made sure that only the tags in "package spec" are checked and all other
+// packages are ignored. This is done because we are expecting the structs to
+// be handled, i.e. marked as "conflicting" only in spec.go, and we do not want
```
Comment (containscafeine) [inline pkg/spec/util.go]: spec.go; since that's where the input structs and the entire spec resides, and all of the other structs are from upstream Kubernetes.
Input structs being from spec.go, and `val.Type().PkgPath()` together make sure that only spec.go is being checked.

File: pkg/spec/util.go [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+
+		// Checking if input is pointer to struct
+		if err := isPointerToStruct(inputStruct); err != nil {
+			return nil, errors.Wrap(err, "Input parameter type mismatch")
+		}
+
+		val := reflect.ValueOf(inputStruct).Elem()
+
+		// Proceeding only if the struct belongs to the kedge's spec package
+		if val.Type().PkgPath() == "github.com/kedgeproject/kedge/pkg/spec" {
```
Comment (containscafeine) [inline pkg/spec/util.go]: well, we could auto generate using `val.Type().PkgPath()`, but that's what we are asserting anyway. For now, I have moved it to the top of the file and made it a const.

File: pkg/spec/util.go [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+			for i := 0; i < val.NumField(); i++ {
+				field := val.Type().Field(i)
+
+				fieldJSONTags := strings.Split(field.Tag.Get("json"), ",")
```
Comment (containscafeine) [inline pkg/spec/util.go]: unsure if we need to use an external library for a simple use case

File: pkg/spec/util.go [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+				// Adding fields that have "conflicting" as a JSON tag to the
+				// array. We need not check the first tag that is used for
```
Comment (containscafeine) [inline pkg/spec/util.go]: umm, we are not making an assumption, we are starting to iterate on the slice from `[1:]` since the first element is not required.
I am making a check and making an actual decision here.

File: pkg/spec/util.go (hunk 2 above)
Comment (containscafeine) [inline pkg/spec/util.go]: hmm, renamed to `checkTypePointerToStruct`

File: pkg/spec/util.go [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+// its underlying type.
+// Returns "nil" if true, or an error if false
+func isPointerToStruct(input interface{}) error {
+	if reflect.Indirect(reflect.ValueOf(input)).Kind().String() != "struct" {
+		return errors.New(fmt.Sprintf("Expected pointer to struct, got %v", reflect.ValueOf(input).Kind().String()))
+	}
+	return nil
+}
+
+// This function takes in pointer to a struct struct as input, and returns a map
```
Comment (containscafeine) [inline pkg/spec/util.go]: fixed

File: pkg/spec/util.go [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+// of conflicting JSON tags used for unmarshaling.
```
Comment (containscafeine) [inline pkg/spec/util.go]: whoops, yep, fixed

File: pkg/spec/util.go [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+// All of the embedded structs are taken into account, and the fields with a
+// JSON tag "conflicting" are assumed to be handled.
+// The returned map contains the JSON tags as the keys, and the values are an
+// array of struct names which contain those fields, without being handled.
+func findConflictingYAMLTags(inputStruct interface{}) (map[string][]string, error) {
```
Comment (containscafeine) [inline pkg/spec/util.go]: fixed

File: pkg/spec/util.go (hunk 9 above)
Comment (surajssd) [inline pkg/spec/util.go]: there is one more comment two lines above that says `package spec`

Comment (surajssd) [inline pkg/spec/util.go]: there is line that says `the tags in "package spec" are checked ` so we should be clear that is what i meant!

**Comment by surajssd** [comment]
soon i will give a second iteration on this!

File: pkg/spec/spec_test.go [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+			if err := checkTypePointerToStruct(inputStruct); err != nil {
+				t.Error(errors.Wrap(err, "Input parameter type mismatch"))
+			}
+
+			conflictingTags, err := findConflictingJSONTags(inputStruct)
+			if err != nil {
+				t.Error(errors.Wrap(err, "Unable to find conflicting tags for spec.App"))
+			}
+			if len(conflictingTags) != 0 {
+				t.Fatalf("The struct %v has unhandled conflicting JSON tags which exist in other structs.\n%v", reflect.Indirect(reflect.ValueOf(inputStruct)).Type().String(), conflictingTags)
```
Comment (surajssd) [inline pkg/spec/spec_test.go]: Try to pretty print what json tag in what all structs! Just dumping that map won't be helpful I think

File: pkg/spec/util.go [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+func getUnmarshalJSONTagsMap(inputMap map[string][]string, inputStruct interface{}, blacklistedTags []string) error {
+
+	// Checking if input is pointer to struct
+	if err := checkTypePointerToStruct(inputStruct); err != nil {
+		return errors.Wrap(err, "Input parameter type mismatch")
+	}
+
+	val := reflect.ValueOf(inputStruct).Elem()
+
+StructFieldsLoop:
```
Comment (surajssd) [inline pkg/spec/util.go]: TIL ! cool stuff

File: pkg/spec/util.go [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+		return nil, errors.Wrapf(err, "Error getting embedded structs for struct: %v", inputStruct)
+	}
+	targetStructs = append(targetStructs, embeddedStructs...)
+
+	// Step 2: Get blacklisted tags.
+	// Get the fields which have already been handled and were previously
+	// conflicting. This needs to be done only for the structs in spec.go,
+	// and with fields marked as "conflicting" using a JSON tag
+
+	blacklistedTags, err := getMarkedAsConflictingJSONUnmarshalTags(targetStructs)
```
Comment (surajssd) [inline pkg/spec/util.go]: if `blacklistedTags` could be map(of type `map[string]interface{}` value does not matter) it will be helpful in searching, and the usefulness of `blacklistedTags` is only to search.

File: pkg/spec/util.go [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+			// first element, e.g. `json:",inline"`
+			continue
+		}
+
+		// Populating only the non blacklisted tags
+		for _, blacklistedTag := range blacklistedTags {
+			if unmarshalJSONTag == blacklistedTag {
+				continue StructFieldsLoop
+			}
+		}
```
Comment (surajssd) [inline pkg/spec/util.go]: if `blacklistedTags` this loop for searching won't be needed and the search will become `O(1)`

File: pkg/spec/spec_test.go [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+		&VolumeClaim{},
+		&ServicePortMod{},
+		&ServiceSpecMod{},
+		&IngressSpecMod{},
+		&EnvFromSource{},
+		&ConfigMapEnvSource{},
+		&Container{},
+		&ConfigMapMod{},
+		&PodSpecMod{},
+		&App{},
```
Comment (surajssd) [inline pkg/spec/spec_test.go]: Correct me if I am wrong, do we need to test everything? I think just testing `&App{}` will test everything recursively? Because in the end everything is embedded in it.

File: pkg/spec/util.go (hunk 9 above)
Comment (containscafeine) [inline pkg/spec/util.go]: ack

File: pkg/spec/util.go [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+
+		inputMap[unmarshalJSONTag] = append(inputMap[unmarshalJSONTag], val.Type().String())
+	}
+	return nil
+}
+
+// This function returns an array of pointers to all of the embedded structs
+// in the input struct.
+// The embedded structs in an embedded struct are also taken into considertation
```
Comment (containscafeine) [inline pkg/spec/util.go]: fixed

File: pkg/spec/util.go [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+				field := val.Type().Field(i)
+
+				fieldJSONTags := strings.Split(field.Tag.Get("json"), ",")
+
+				// Adding fields that have "conflicting" as a JSON tag to the
+				// array. We need not check the first tag that is used for
+				// unmarshalling.
+				for _, tag := range fieldJSONTags[1:] {
+					if tag == "conflicting" {
+						blacklistedTags = append(blacklistedTags, fieldJSONTags[0])
```
Comment (containscafeine) [inline pkg/spec/util.go]: fixed

File: pkg/spec/util.go [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+	}
+	return blacklistedTags, nil
+}
+
+// This function checks if the input parameter has the pointer to a struct as
+// its underlying type.
+// Returns "nil" if true, or an error if false
+func checkTypePointerToStruct(input interface{}) error {
+	if reflect.Indirect(reflect.ValueOf(input)).Kind().String() != "struct" {
+		return errors.New(fmt.Sprintf("Expected pointer to struct, got %v", reflect.ValueOf(input).Kind().String()))
```
Comment (containscafeine) [inline pkg/spec/util.go]: fixed

File: pkg/spec/util.go [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+func findConflictingJSONTags(inputStruct interface{}) (map[string][]string, error) {
+
+	// We need to check that no JSON tag is specified more than once at the
+	// top level of the struct.
+	// This means that the struct fields, and the JSON tags of all the
+	// fields of all the embedded structs need to be unique.
+	// We accomplish this over a 4 step process.
+
+	// Checking if input is pointer to struct
+	if err := checkTypePointerToStruct(inputStruct); err != nil {
```
Comment (containscafeine) [inline pkg/spec/util.go]: I'd ideally want to call this in every function that is taking in an `interface{}` to check if the input value type is what we want it to be. Makes it more testable and more robust, imo. We never know where what gets used when, and I'd want the function to be complete in itself.

File: pkg/spec/util.go (hunk 18 above)
Comment (containscafeine) [inline pkg/spec/util.go]: PTAL at https://www.darkcoding.net/software/go-slice-search-vs-map-lookup/
Also, during Kennedy's workshop he demonstrated that underlying arrays are always going to be much faster than any other data structure for most cases.

File: pkg/spec/spec_test.go (hunk 20 above)
Comment (containscafeine) [inline pkg/spec/spec_test.go]: Yep, the embedded structs will be taken care of, but not -
```golang
	VolumeClaims               []VolumeClaim     `json:"volumeClaims,omitempty"`
	ConfigMaps                 []ConfigMapMod    `json:"configMaps,omitempty"`
	Services                   []ServiceSpecMod  `json:"services,omitempty"`
	Ingresses                  []IngressSpecMod  `json:"ingresses,omitempty"`
```, but we also need to test those.

Ideally we should be getting all the structs in this file and test all one by one, and ignore testing if tested before, using ast or something, but that's for another PR sometime I guess.

File: pkg/spec/spec_test.go [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+	}
+
+	for _, inputStruct := range structsToTest {
+		t.Run("Testing conflicting fields", func(t *testing.T) {
+
+			// Checking if input is pointer to struct
+			if err := checkTypePointerToStruct(inputStruct); err != nil {
```
Comment (containscafeine) [inline pkg/spec/spec_test.go]: answered above

File: pkg/spec/spec_test.go (hunk 16 above)
Comment (containscafeine) [inline pkg/spec/spec_test.go]: The map has the information about which tag exists in which struct(s), for me printing it was enough.
Can you try to make it fail and see it the debug message is not clear enough for you?

File: pkg/spec/util.go (hunk 24 above)
Comment (surajssd) [inline pkg/spec/util.go]: I think we should add things when we need it, if we are not sure about how it is gonna be used. Let's not commit for future but do what is needed now.

But for what is being done right now, I don't see it's use!

File: pkg/spec/util.go (hunk 18 above)
Comment (surajssd) [inline pkg/spec/util.go]: the blog says:
>tl;dr Use a map.

File: pkg/spec/spec_test.go (hunk 20 above)
Comment (surajssd) [inline pkg/spec/spec_test.go]: Ahh, ok

File: pkg/spec/spec_test.go (hunk 16 above)
Comment (surajssd) [inline pkg/spec/spec_test.go]: Yep I tried, that's why suggesting that change.

File: pkg/spec/util.go (hunk 18 above)
Comment (containscafeine) [inline pkg/spec/util.go]: Yes, that is for > 10 items. I don't see us handling > 10 conflicting fields in the next year :)

File: pkg/spec/util.go (hunk 24 above)
Comment (containscafeine) [inline pkg/spec/util.go]: Not sure :( This is not making it future proof, it's making every function complete. Maybe look at it like making every input parameter type safe. Since we are taking an interface{} as an input which can take literally any type, we _must_ make sure that the type is what we want.

Comment (surajssd) [inline pkg/spec/util.go]: then those functions should return specific types than returning interfaces, I think because doing check once is enough and extracting the concrete type should be done there.

Comment (surajssd) [inline pkg/spec/util.go]: then those functions should return specific types than returning interfaces, I think because doing check once is enough and extracting the concrete type should be done there.

Comment (surajssd) [inline pkg/spec/util.go]: then those functions should return specific types than returning interfaces, I think because doing check once is enough and extracting the concrete type should be done there.

File: pkg/spec/util.go (hunk 18 above)
Comment (surajssd) [inline pkg/spec/util.go]: cool, let's not do something we don't need right now!

File: pkg/spec/util.go (hunk 24 above)
Comment (containscafeine) [inline pkg/spec/util.go]: That would be ideal, but we cannot return specific concrete types here, e.g. we cannot return `*PodSpecMod` because we have to pass all the different structs in spec.go around e.g. `*ServiceSpecMod`, `*App`, etc. The only way way I could find was passing interfaces and making sure that passed values are type safe later on and making every function complete. What do you suggest?
ping @kadel could you help here as to what would be the right way forward?

Comment (containscafeine) [inline pkg/spec/util.go]: After a meeting with @surajssd, we decided that instead of passing interface{} around which can be easily abused, I'll be refactoring this to pass `reflect.Value` as the concrete type to all functions.

File: pkg/spec/spec_test.go (hunk 16 above)
Comment (containscafeine) [inline pkg/spec/spec_test.go]: Sure, I've changed it to the following -
```console
$ go test ./pkg/...
?       github.com/kedgeproject/kedge/pkg/cmd   [no test files]
ok      github.com/kedgeproject/kedge/pkg/encoding      0.095s
--- FAIL: TestConflictingFields (0.00s)
    --- FAIL: TestConflictingFields/Testing_conflicting_fields#06 (0.00s)
        spec_test.go:40: The struct spec.Container has unhandled conflicting JSON tags which exist in other structs.
        spec_test.go:42: The JSON tag 'envFrom' exists in [spec.Container v1.Container]
        spec_test.go:44: Once you handle the above conflicting JSON tag, mark it as handled by adding a 'conflicting' JSON tag to its definition. e.g.
                Containers     []Container `json:"containers,conflicting,omitempty"`
FAIL
FAIL    github.com/kedgeproject/kedge/pkg/spec  0.052s
ok      github.com/kedgeproject/kedge/pkg/transform/kubernetes  0.048s
```

**Comment by containscafeine** [comment]
@surajssd I've removed passing around `interface{}` and moved to `reflect.Value`. Also, refactored the failing test output. PTAL!

**Comment by kadel** [comment]
Code looks good, and it seems to be working ;-)

Just one question. All the functions that you created in `pkg/spec/util.go` are part of the tests. Wouldn't it be better to have them in `*_test.go` file/files? Those functions are for testing and they are not part of the  " main program"

**Comment by containscafeine** [comment]
@kadel hmm, makes sense.
I've made the following changes, let me know if you have better suggestions for file names :(
```console
        renamed:    util.go -> conflicting_fields_test.go
        renamed:    util_test.go -> conflicting_fields_util_test.go
```

**Comment by kadel** [comment]
It might be even more confusing now than it was before :-D 

Do I understand it correctly that  `conflicting_fields_util_test.go` is test for tests? :-D 


Can we have all the tests in `spec_test.go`? I think it would make much more sense. Especially with Go conventino keeping all tests testing `foo.go` functionality in `foo_test.go` file.

**Comment by containscafeine** [comment]
@kadel makes sense, changed to
```console
renamed:    pkg/spec/conflicting_fields_util_test.go -> pkg/spec/conflicting_fields_test_test.go
```

> Can we have all the tests in spec_test.go? I think it would make much more sense. Especially with Go conventino keeping all tests testing foo.go functionality in foo_test.go file.

But that means we are testing stuff in spec.go, instead we are testing stuff in conflicting_fields_test.go.

**Comment by kadel** [comment]
> But that means we are testing stuff in spec.go, instead we are testing stuff in conflicting_fields_test.go.

I meant that right place for `conflicting_fields_test.go` should be in `spec_test.go` no?

**Comment by kadel** [comment]
Maybe it doesn't matter, and I'm just complicating this too much :blush: :innocent:

**Comment by containscafeine** [comment]
@kadel makes sense, I've moved everything from `conflicting_fields_test.go` to `spec_test.go`, and renamed `conflicting_fields_test_test.go` to `conflicting_fields_test.go`.

**Comment by containscafeine** [comment]
squashin' & mergin' :fire:
